File: Stacks_Queues/go/MyQueue.go
package main
import (
  // "fmt"
  "types"
)


type MyQueue struct {
  queue *types.Stack
  buffer *types.Stack
}


func (q *MyQueue) Initialize() {
  q.queue = &types.Stack{}
  q.buffer = &types.Stack{}
}


func (q *MyQueue) Add(x int) {
  q.queue.Push(x)
}


func (q *MyQueue) Remove() (int, bool) {
  if q.IsEmpty() {
    return 0, false
  }
  var top, value int

  for !q.IsEmpty() {
    top, _ = q.queue.Pop()
    q.buffer.Push(top)
  }

  value, _ = q.buffer.Pop()

  for !q.buffer.IsEmpty() {
    top, _ = q.buffer.Pop()
    q.queue.Push(top)
  }

  return value, true
}


func (q *MyQueue) IsEmpty() bool {
  return q.queue.IsEmpty()
}

func PrintQueue(q MyQueue) {
  types.PrintStack(*q.queue)
}


func main() {
  var q MyQueue
  q.Initialize()

  q.Add(2)
  q.Add(5)
  q.Add(3)
  q.Add(1)
  PrintQueue(q)
  q.Remove()
  q.Remove()
  PrintQueue(q)
}

File: Stacks_Queues/python/MyQueue.py
from Stack import Stack
from exceptions import QueueEmptyException


class MyQueue(object):
    def __init__(self):
        self.queue = Stack()
        self.buffer = Stack()

    def add(self, x):
        self.queue.push(x)

    def remove(self):
        if self.is_empty():
            raise QueueEmptyException()

        counter = 0
        while not self.queue.is_empty():
            self.buffer.push(self.queue.pop())
            counter += 1

        value = self.buffer.pop()

        while not self.buffer.is_empty():
            self.queue.push(self.buffer.pop())

        return value

    def is_empty(self):
        return self.queue.is_empty()

    def __str__(self):
        return self.queue.__str__()


q = MyQueue()
q.add(3)
q.add(4)
q.add(5)
q.add(10)
print(q)
print(q.remove())
print(q.remove())
q.add(101)
print(q)
